apps/service_providers/llm_service/runnables.py

```diff
@@ -267,25 +272,129 @@ def invoke(
 
         if config.get("configurable", {}).get("save_input_to_history", True):
             chat_message = self.state.save_message_to_history(input, ChatMessageType.HUMAN)
-            if resource_file_ids:
-                chat_message.metadata = chat_message.metadata | resource_file_ids
-                chat_message.save()
+            if human_message_resource_file_ids:
+                self._add_resource_files_to_message_metadata(chat_message, human_message_resource_file_ids)
 
         # Note: if this is not a new chat then the history won't be persisted to the thread
         thread_id = self.state.get_metadata(Chat.MetadataKeys.OPENAI_THREAD_ID)
         if thread_id:
             input_dict["thread_id"] = thread_id
 
         input_dict["instructions"] = self.state.get_assistant_instructions()
-
         response = self._get_response_with_retries(config, input_dict, thread_id)
+
+        output, ai_message_resource_file_ids = self._save_response_annotations(response)
+
         if not thread_id:
             self.state.set_metadata(Chat.MetadataKeys.OPENAI_THREAD_ID, response.thread_id)
 
-        output = response.return_values["output"]
-        self.state.save_message_to_history(output, ChatMessageType.AI)
+        ai_message = self.state.save_message_to_history(output, ChatMessageType.AI)
+        if ai_message_resource_file_ids:
+            self._add_resource_files_to_message_metadata(ai_message, ai_message_resource_file_ids)
+
+        return ChainOutput(output=output, prompt_tokens=0, completion_tokens=0)
+
+    @transaction.atomic()
+    def _save_response_annotations(self, response: OpenAIAssistantFinish) -> tuple[str, dict]:
```

Comment:
I think this can just return a list of `file_ids` now since we aren't using the other info outside of this function.

Reply:
https://github.com/dimagi/open-chat-studio/pull/545/commits/defd877b67e088c5d20b528702e478d473ea88b2